### PR TITLE
Feature/budget-calculator

### PR DESCRIPTION
**Pull request by TheaAleksovDev**

### CODE REVIEW

**Comment by valentin-aleksandrov** [comment]
Can you update your commit messages according to the guidelines we discussed in other PRs?

File: src/app/settings/settings.component.css [hunk 1]
```diff
@@ -0,0 +1,34 @@
+button {
+  padding: 10px 15px;
+  box-sizing: border-box;
+  border: none;
+  color: #45a049;
+  /* background-color: #99e89d; */
+
```
Comment (valentin-aleksandrov) [inline src/app/settings/settings.component.css]:
```suggestion
```

File: src/app/settings/settings.component.html [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
-  </button>
-  <button (click)="setRole('guest')" [class.selected]="role() === 'guest'">
-    Guest
+  @for(element of allRoles; track element.role){
+  <button
+    (click)="setRole(element.role)"
+    [class.selected]="role() === element.role"
+    [disabled]="!element.isAvailable && true"
+  >
+    {{ element.role.toUpperCase() }} {{ !element.isAvailable ? "X" : "" }}
```
Comment (valentin-aleksandrov) [inline src/app/settings/settings.component.html:19]: Do we need the **!** operator?
Can you reverse the conditions and remove the **!** operator?

File: src/app/settings/settings.component.html [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<div class="choose">
+  @for(element of allCurrencies; track element.currency){
+  <button
+    (click)="setCurrency(element.currency)"
+    [class.selected]="currency() === element.currency"
+    [disabled]="!element.isAvailable && true"
+  >
+    {{ element.currency.toUpperCase() }} {{ !element.isAvailable ? "X" : "" }}
```
Comment (valentin-aleksandrov) [inline src/app/settings/settings.component.html]: Do we need the ! operator?
Can you reverse the conditions and remove the ! operator?

File: src/app/settings/settings.component.html [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  }
+</div>
+<div class="choose">
+  @for(element of allRoles; track element.role){
+  <button
+    (click)="setRole(element.role)"
+    [class.selected]="role() === element.role"
+    [disabled]="!element.isAvailable && true"
+  >
+    {{ element.role.toUpperCase() }} {{ !element.isAvailable ? "X" : "" }}
```
Comment (valentin-aleksandrov) [inline src/app/settings/settings.component.html]: Same as my comment on line **8**.

**Comment by valentin-aleksandrov** [comment]
image

It's great that you are separating the bug fixes into separate commits.

I would challenge you to describe those fixes.
What is being fixed - the less I need to open commits, the less time I spend on such things that could have been explained with a more detailed message.
*fix a bug* is too generic and it doesn't give me as much details as I would hope to get.

File: src/app/currency.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+export class CurrencyConvertPipe implements PipeTransform {
+  transform(value: number, currency: Currency, convertToCurrency: Currency) {
+    if (!value || isNaN(value)) {
+      return null;
+    }
+    // console.log(value, currency, convertToCurrency);
+    if (currency === convertToCurrency) {
+      return value.toFixed(2);
+    } else if (currency === 'USD' && convertToCurrency === 'BGN') {
+      return (value * 1.75).toFixed(2);
```
Comment (valentin-aleksandrov) [inline src/app/currency.pipe.ts]: Numbers such as `1.75` and `0.57` can be extracted into a separate file called:
- `conversion-rates.ts`

```ts
export const USD_TO_BGN_CONVERSATION_RATE = 1.75
export const BGN_TO_USD_CONVERSION_RATE = 0.57
```
This will increase code readability.

@TheaAleksovDev - would you be interested If I were to create a new issue to enhance the Application to fetch the rates from an REST API? This will be a new task, where you will implement the [REST polling pattern](https://medium.com/weekly-webtips/polling-in-angular-how-to-use-rxjs-in-angular-for-polling-14c519f4c218).
Basically fetching currency rates every minute or SO.

File: src/app/highlight-by-amount.directive.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  updateBorderColor = () => {
+    if (this.amount() > 0) {
+      if (this.amount() < 500) {
+        this.borderColor.set('#5bff63');
+      } else if (this.amount() > 500 && this.amount() < 1000) {
+        this.borderColor.set('#45a049');
+      } else {
+        this.borderColor.set('#295f2c');
```
Comment (valentin-aleksandrov) [inline src/app/highlight-by-amount.directive.ts:36]: If `this.amount() === 500` what should happen?

File: src/app/highlight-by-amount.directive.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      }
+    } else {
+      if (this.amount() > -500) {
+        this.borderColor.set('#d70000');
+      } else if (this.amount() < -500 && this.amount() > -1000) {
+        this.borderColor.set('#b80000');
+      } else if (this.amount() < -1000) {
```
Comment (valentin-aleksandrov) [inline src/app/highlight-by-amount.directive.ts:43]: If `this.amount() === -500` what should happen?
In the case of `this.amount() === -1000` what is the expectation?

File: src/app/currency.pipe.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+@Pipe({
+  name: 'convert',
+  standalone: true,
+})
+export class CurrencyConvertPipe implements PipeTransform {
+  transform(value: number, currency: Currency, convertToCurrency: Currency) {
+    if (!value || isNaN(value)) {
+      return null;
+    }
+    // console.log(value, currency, convertToCurrency);
```
Comment (valentin-aleksandrov) [inline src/app/currency.pipe.ts]:
```suggestion
```

File: src/app/components/budget-inputs/balance.service.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    ]);
+    this.balance.update((balance) => balance + amount);
+    this.saveBalance();
+    this.saveTransactions();
+  };
+
+  saveBalance = () => {
+    localStorage.setItem('balance', this.balance().toString());
+  };
+  saveTransactions = () => {
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts:86]: I really like that you are using some persistent layer. When I'm reviewing your PRs it really helps me to focus on the code and not on the setup.
Good Job, Tea!

File: src/app/components/budget-inputs/budget-inputs.component.html [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  balance:
+  {{ (balance().amount | convert : balance().currency : currency()) || "0" }}
+  {{ currency() }}
+</h2>
+<p class="info" *appShowByRole="role()">
+  *Balance might vary due to the currency exchange rate and mixed values*
+</p>
+<div class="inputs" *appShowByRole="role()">
+  <input
+    type="text"
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/budget-inputs.component.html]:
```suggestion
    type="number"
```

File: src/app/components/budget-display/budget-display.component.html [hunk 11]
```diff
@@ -0,0 +1,36 @@
+<div class="transactions">
+  <div class="incomes">
+    @for(income of incomes(); track income){
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-display/budget-display.component.html]: The `track` is to specify a unique property from the income object such as an id for example.
I haven't found on the internet an example where developers are passing the object itself - it is also something `unique`. The reason I couldn't find such example is most probably because when you use the object, its references could change and that could potentially could trick Angular into retendering.

In React we have something similar - the `key`
```jsx
[{ id: 1, name: 'a' }, { id: 2, name: 'b' }, { id: 3, name: 'a' }]
   .map(element => (<div key={element.id}>{element.name}</div>))  
```

File: src/app/components/budget-inputs/balance.service.ts (hunk 9 above)
Comment (TheaAleksovDev) [reply src/app/components/budget-inputs/balance.service.ts:86]: I'm glad that it makes the reviewing process easier for you.

File: src/app/currency.pipe.ts (hunk 5 above)
Comment (TheaAleksovDev) [reply src/app/currency.pipe.ts]: That sounds interesting. Yes, I'd like another challenge.

File: src/app/highlight-by-amount.directive.ts (hunk 6 above)
Comment (TheaAleksovDev) [reply src/app/highlight-by-amount.directive.ts:36]: I'll resolve this in the hover issue that you opened.

File: src/app/highlight-by-amount.directive.ts (hunk 7 above)
Comment (TheaAleksovDev) [reply src/app/highlight-by-amount.directive.ts:43]: Same as my reply to the previous comment

**Comment by TheaAleksovDev** [comment]
I am aware that "resolve comments" is too generic name, but should I really list everything that I did in that commit?

**Comment by TheaAleksovDev** [comment]
> image
> 
> It's great that you are separating the bug fixes into separate commits.
> 
> I would challenge you to describe those fixes. What is being fixed - the less I need to open commits, the less time I spend on such things that could have been explained with a more detailed message. _fix a bug_ is too generic and it doesn't give me as much details as I would hope to get.

I understand. From now on, I'll make these commits more specific.

File: src/app/settings/settings.component.css [hunk 12]
```diff
@@ -0,0 +1,33 @@
+button {
+  padding: 10px 15px;
+  box-sizing: border-box;
+  border: none;
+  color: #45a049;
+
```
Comment (valentin-aleksandrov) [inline src/app/settings/settings.component.css]: **Optional**:
```suggestion
```

File: src/app/currency.pipe.ts (hunk 5 above)
Comment (valentin-aleksandrov) [reply src/app/currency.pipe.ts]: nice

Here you go:
- https://github.com/TheaAleksovDev/budgetCalculator/issues/6

**Comment by valentin-aleksandrov** [comment]
> I am aware that "resolve comments" is too generic name, but should I really list everything that I did in that commit?

"resolve comments" is fine in most of the cases.

If some of the changes are something bigger or you think is better to have it in a separate commit - feel free to do that as well.

File: src/app/components/budget-inputs/balance.service.ts [hunk 13]
```diff
@@ -28,24 +33,49 @@ export class BalanceService {
     return this.expenses();
   };
 
-  addIncome = (amount: number, description: string) => {
+  updateBalance(currency: Currency, amount: number, operator: '+' | '-') {
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts:30]: I like how you have typed the `operator` parameter - good job!

File: src/app/app.component.css [hunk 14]
```diff
@@ -0,0 +1,8 @@
+.app{
```
Comment (valentin-aleksandrov) [inline src/app/app.component.css]: Can you format this file?

File: src/app/components/budget-inputs/balance.service.ts [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+    this.updateBalance(currency, amount, '+');
+  };
+
+  addExpense = (amount: number, currency: Currency, description: string) => {
+    const id = nanoid();
+    this.expenses.update((incomes) => [
+      ...incomes,
+      {
+        amount: amount,
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts]:
```suggestion
        amount,
```

File: src/app/components/budget-inputs/balance.service.ts [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        currency: currency,
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts]:
```suggestion
        currency,
```

File: src/app/components/budget-inputs/balance.service.ts [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        description: description,
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts]:
```suggestion
        description,
```

File: src/app/components/budget-inputs/balance.service.ts [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+        id: id,
```
Comment (valentin-aleksandrov) [inline src/app/components/budget-inputs/balance.service.ts]:
```suggestion
        id,
```

**Review by valentin-aleksandrov — APPROVED** [review]
